[Futures] Added markCompletedIfNeeded API

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39114 [Future] modularized running callbacks on future completion
* #39113 [Futures] Allow setErrorIfNeeded arg to have type FutureError
* **#39080 [Futures] Added markCompletedIfNeeded API**

This PR adds a function similar to setErrorIfNeeded for marking
futures complete. It only completes futures if they haven't been completed
already.

Fixes: https://github.com/pytorch/pytorch/issues/39082

Differential Revision: [D21746065](https://our.internmc.facebook.com/intern/diff/D21746065/)

osalpekar committed:
[Futures] Added markCompletedIfNeeded API

This PR adds a function similar to setErrorIfNeeded for marking
futures complete. It only completes futures if they haven't been completed
already.

Differential Revision: [D21746065](https://our.internmc.facebook.com/intern/diff/D21746065/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/utils/future.h:
```diff
@@ -156,6 +154,25 @@ class TORCH_API Future final {
+  void markCompletedInternal(T value,
+      std::unique_lock<std::mutex>& lock) {
+    TORCH_CHECK(!completed_);
+    // Set value first as completed_ is accessed without lock
```
What does this comment mean? Looks like the lock is always held before entering this method, no?

osalpekar replied:
It seems incorrect bc we have the lock - the comment was just a remnant of the old markCompleted function, but I'll remove

mrshenli reviewed (approved):
LGTM! Thanks for adding.

osalpekar committed:
Update on "[Futures] Added markCompletedIfNeeded API"


This PR adds a function similar to setErrorIfNeeded for marking
futures complete. It only completes futures if they haven't been completed
already.

Fixes: https://github.com/pytorch/pytorch/issues/39082

Differential Revision: [D21746065](https://our.internmc.facebook.com/intern/diff/D21746065/)

[ghstack-poisoned]

jjlilley commented on torch/csrc/utils/future.h:
```diff
@@ -61,25 +61,23 @@ class TORCH_API Future final {
-    for (auto& callback : cbs) {
-      callback();
+    if (completed_) {
+      LOG(INFO) << "Skipping marking the Future completed since it has " <<
```
Do we really want to log?
I realize that `setErrorIfNeeded()` logs, though we might log if the caller wants it by returning true/false whether we were able to markCompleted.
 
(Or, if we do log, can we at least not call operator<<() twice?)

osalpekar committed:
Update on "[Futures] Added markCompletedIfNeeded API"


This PR adds a function similar to setErrorIfNeeded for marking
futures complete. It only completes futures if they haven't been completed
already.

Fixes: https://github.com/pytorch/pytorch/issues/39082

Differential Revision: [D21746065](https://our.internmc.facebook.com/intern/diff/D21746065/)

[ghstack-poisoned]

xush6528 commented on torch/csrc/utils/future.h:
```diff
@@ -156,6 +153,24 @@ class TORCH_API Future final {
   }
 
+  void markCompletedInternal(T value,
+      std::unique_lock<std::mutex>& lock) {
```
It takes a reference, while you are passing rvalue with `std::move`. 

Do you want to move a rvalue instead of passing reference?

`std::unique_lock` should be movable.

Change to `std::unique_lock<std::mutex> lock`?

osalpekar committed:
Update on "[Futures] Added markCompletedIfNeeded API"


This PR adds a function similar to setErrorIfNeeded for marking
futures complete. It only completes futures if they haven't been completed
already.

Fixes: https://github.com/pytorch/pytorch/issues/39082

Differential Revision: [D21746065](https://our.internmc.facebook.com/intern/diff/D21746065/)

[ghstack-poisoned]
